README.md
# Uwt

Uwt provides OCaml bindings
for [libuv](https://github.com/libuv/libuv) - on top
of [lwt](https://github.com/ocsigen/lwt).

[![Travis build Status](https://travis-ci.org/fdopen/uwt.svg?branch=master)](https://travis-ci.org/fdopen/uwt)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/fdopen/uwt?svg=true)](https://ci.appveyor.com/project/fdopen/uwt)

* Requests are translated to lwt-threads.

* Callbacks that are called continually are most of the time not
  translated to the usual lwt semantic.

* naming conventions mirror the conventions of libuv, so you can easily
  consult the official [libuv manual](http://docs.libuv.org/en/v1.x/).
  Only the differences are explained inside `uwt.mli`

* Uwt is **not** compatible with `lwt.unix`. It's not a further
  `Lwt_engine` in addition to `select` and `libev`. However, the
  findlib package `uwt.ext` provides alternatives to the higher level
  modules of `lwt.unix` with a nearly identic interface (Uwt_io,
  Uwt_process, Uwt_log). You can also open the module `Uwt_compat`
  (package `uwt.compat`) that provides aliases with module names as
  found in lwt (e.g. `Lwt_main` instead of `Uwt.Main`). This is
  necessary, if you want to use the camlp4 syntax extension of lwt.

* Uwt is not thread safe. All uwt functions should be called from your
  main thread.

## Installation

### Quick install guide

```bash
$ opam install uwt
# or for the latest development version
$ opam pin add uwt --dev-repo
```

### Manual install

Dependencies:

* OCaml 4.02.1 or later
* lwt 2.6.0 or later
* libuv 1.8 or later

Build dependencies:

* pkg-config / pkgconf
* findlib
* omake
* cppo
* ppx_deriving (test only)
* ppx_import (test only)
* ounit (test only)
* autoconf (repo pinning only)

```bash
$ omake all
$ omake install
```

libuv will be compiled locally, if is not already installed on your system.
You can also explicity enforce one option:

```bash
$ omake all BUILD_LIBUV=true
# or
$ omake all BUILD_LIBUV=false
```


CHANGES.md
0.3.3 2018-08-24
---------------------------

- update to libuv 1.23.0:
   * new functions: `Uwt.Misc.setpriority`, `Uwt.Misc.getpriority`,
     and `Uwt.Misc.getpid`. (`Unix.getpid` doesn't return
     `GetCurrentProcessID` under Windows ...)

- fix compilation

0.3.2 2018-07-16
---------------------------

- update to libuv 1.22.0:
   * new function: `Uwt.Fs.lchmown`

- Change license to MIT

- fix compilation under OCaml 4.07.0 (Windows)

- add new `Uwt.error` codes: `EHOSTDOWN`, `EREMOTEIO`, `ENOTTY`, and
  `EFTYPE`

0.3.0.2 2018-04-23
---------------------------

Maintenance release, because the bundled libuv version of the previous
release contains a serious regression.

- update to libuv 1.20.2

- File related functions inside `Uwt_compat.Lwt_io` now take
  parameters of type `Unix.open_flag` instead of `Uwt.Fs.uv_open_flag`
  
0.3.0 2018-04-15
---------------------------

- update to lwt 4.0.0:
   * The modules `Uwt_chan` and `Uwt_log` have been removed. `Uwt_log`
     is available as an external library
   * `Uwt_io.file_length` now fails with `EISDIR` when used on a
     directory (see https://github.com/ocsigen/lwt/issues/563)

- update to libuv 1.20.0:
   * `Uwt.Fs.copyfile` now supports file cloning

0.2.4 2018-02-10
---------------------------

- fix compilation under OCaml 4.06.0 (Windows)

- update to libuv 1.19.1:
   * new flags for `Uwt.Fs.openfile`: `O_DIRECT`, `O_EXLOCK`,
     `O_NOATIME`, `O_SYMLINK`, `O_NOFOLLOW`, and `O_DIRECTORY`
   * new function: `Uwt.Pipe.chmod` to allow access from other users
   * new function: `Uwt.Misc.getppid` (similar to `Unix.getppid`, but also
     supported under Windows).

0.2.3 2017-09-08
---------------------------

- update to libuv 1.14.1

- new functions from `Lwt_io`: `Uwt_io.open_temp_file` and
  `Uwt_io.with_temp_file`

- `Uwt_base.Sys_info.win_version` now reports your current Windows
  version, even if your executable is not manifested.

0.2.2 2017-08-17
---------------------------

- update to libuv 1.14.0:
  * new function: `Uwt.Fs.copyfile`
  * `Uwt.Poll.start` now supports watching for sysfs interrupts or TCP
    out-of-band messages
  * `Uwt.Udp.try_send` now works under Windows

- `Uwt.Tcp.enable_keepalive` will now fail, if you pass an `Uwt.Tcp.t`
  handle to it, that doesn't wrap a socket yet. libuv silently ignores
  the delay parameter in this case.  You can use `Uwt.Tcp.init_ipv4` or
  `Uwt.Tcp.init_ipv6` instead of `Uwt.Tcp.init` to circumvent this problem.

- `Uwt_io` and `Uwt_log` have been updated to match the recent `lwt.unix`
  changes.

0.2.1 2017-06-01
---------------------------

- update internal libuv version to to 1.12.0
  * new functions: `Uwt.Signal.oneshot` and
    `Uwt_base.Misc.{put,set,unset}env` (added because of utf8 support
    under windows)

0.2.0 2017-12-05
---------------------------
- `Uwt.Misc.guess_handle` now takes an `Unix.file_descr` as parameter
   (instead of `Uwt.file`). This doesn't follow the libuv prototypes,
   but makes more sense in the OCaml context. `Uwt.Misc.guess_handle`
   can also classify UDP and TCP sockets on Windows.

- The IPC interface has changed:
  * `Uwt.Stream.write2` has been removed. There are now properly typed
     `Uwt.Pipe.write2*` functions that can also send UDP handles to a
     child process.
  * `Uwt.Stream.accept_raw` has been removed. Instead
    `Uwt.Pipe.accept_ipc` must be used to accept handles that were
    send via the libuv specific IPC interface. `Uwt.Tcp.accept_raw`
    and `Uwt.Pipe.accept_raw` still exist.
  * The `Uwt.Process.stdio` type has been
    extended. `Uwt.Process.spawn` now allows to create duplex stream
    for IPC (previously not exposed by uwt)

 - Exception update: Many functions `Lwt.fail` with `exception
   Unix_error of error * string * string` (or the `foo_exn` functions
   will raise such exceptions). The first string component of this
   exception is supposed to contain the function name. However there
   was no consistency what was put there as name. If the function is
   called uv_foo (or `uv_xy_foo`, `uv_foo_xy`, etc.) and `man 2 foo`
   or `man 3 foo` lists a closely related function that could be
   wrapped by `uv_foo`, `foo` is used as function name (even if a
   different, system dependent function is used internally by
   libuv). This leads to more consistent behaviour regarding to `Unix`
   and `Lwt_unix`. If it's not obvious (by just looking at the name),
   which function is called by libuv, `uv_foo` is used instead.

- new functions: `Uwt.Fs.pread` / `Uwt.Fs.pwrite` / `Uwt.Fs.pwritev`

- The mli files are now better documented.

- `Uwt.Tcp.keepalive` has been split in two functions:
  `Uwt.Tcp.enable_keepalive` and `Uwt.Tcp.disable_keepalive`

- Tasks that are intended to run in libuv's threadpool and are
  canceled via `Lwt.cancel` now report the success of the cancellation
  immediately instead of slightly deferred. So the following check
  will work now:

  ```ocaml
  let s = Uwt.Fs.stat "foo" in
  (* more code *)
  assert (Lwt.state s = Lwt.Sleep); (* relevant use case *)
  Lwt.cancel s;
  match Lwt.state s with
  | Lwt.Sleep -> print_endline "cancel failed, the job is already running in background"
  | Lwt.Fail Lwt.Canceled -> print_endline "cancel success"
  | _ -> prerr_endline "Huh?"```

- Bugs in the following functions have been fixed:
  * `Uwt.Udp.set_multicast_ttl` (it rejected valid parameters)
  * `Uwt.Handle_ext.get_send_buffer_size` /
    `Uwt.Handle_ext.get_recv_buffer_size` (they've returned wrong
    results)
  * `Uwt.Stream.write2` (now ``Uwt.Pipe.write2`) (invalid memory
    access under Windows, because libuv's interface slightly differs
    on windows and unix)

0.1.0 2017-03-02
---------------------------
- new functions: Uwt.Stream.writev and Uwt.Fs.writev

- the opam package will now always use the internal copy of
  libuv.

- prepare OCaml 4.05 support

- internal libuv version updated to 1.11.0

- avoid memcpy calls inside Uwt.Stream.read and Uwt.Udp.recv
  (*nix only)


src/uwt_log.ml
(* Lightweight thread library for OCaml
 * http://www.ocsigen.org/lwt
 * Module Lwt_log
 * Copyright (C) 2002 Shawn Wagner <raevnos@pennmush.org>
 *               2009 Jérémie Dimino <jeremie@dimino.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, with linking exceptions;
 * either version 2.1 of the License, or (at your option) any later
 * version. See COPYING file for details.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 *)

(* This code is an adaptation of [syslog-ocaml] *)

include Lwt_log_core

open Lwt.Infix

let program_name = Filename.basename Sys.executable_name

(* Errors happening in this module are always logged to [stderr]: *)
let log_intern fmt =
  Printf.eprintf ("%s: Lwt_log: " ^^ fmt ^^ "\n%!") program_name

(* +-----------------------------------------------------------------+
   | Templates                                                       |
   +-----------------------------------------------------------------+ *)

let date_string time =
  let tm = Unix.localtime time in
  let month_string =
    match tm.Unix.tm_mon with
      | 0 -> "Jan"
      | 1 -> "Feb"
      | 2 -> "Mar"
      | 3 -> "Apr"
      | 4 -> "May"
      | 5 -> "Jun"
      | 6 -> "Jul"
      | 7 -> "Aug"
      | 8 -> "Sep"
      | 9 -> "Oct"
      | 10 -> "Nov"
      | 11 -> "Dec"
      | _ -> Printf.ksprintf failwith "Lwt_log.ascdate: invalid month, %d" tm.Unix.tm_mon
  in
  Printf.sprintf "%s %2d %02d:%02d:%02d" month_string tm.Unix.tm_mday tm.Unix.tm_hour tm.Unix.tm_min tm.Unix.tm_sec

let render ~buffer ~template ~section ~level ~message =
  let time = lazy(Unix.gettimeofday ()) in
  let file, line, column =
    match Lwt.get location_key with
      | Some loc -> loc
      | None -> ("<unknown>", -1, -1)
  in
  Buffer.add_substitute buffer
    (function
       | "date" -> date_string (Lazy.force time)
       | "milliseconds" -> Printf.sprintf "%03.0f" (mod_float (Lazy.force time *. 1000.) 1000.)
       | "name" -> program_name
       | "pid" -> string_of_int (Unix.getpid ())
       | "message" -> message
       | "level" -> Lwt_log_core.string_of_level level
       | "section" -> Section.name section
       | "loc-file" -> file
       | "loc-line" -> string_of_int line
       | "loc-column" -> string_of_int column
       | var -> Printf.ksprintf invalid_arg "Lwt_log.render_buffer: unknown variable %S" var)
    template

(* +-----------------------------------------------------------------+
   | Predefined loggers                                              |
   +-----------------------------------------------------------------+ *)

let channel ?(template="$(name): $(section): $(message)") ~close_mode ~channel () =
  make
    ~output:(fun section level lines ->
               Uwt_io.atomic begin fun oc ->
                 let buf = Buffer.create 42 in
                 Lwt_list.iter_s
                   (fun line ->
                      Buffer.clear buf;
                      render ~buffer:buf ~template ~section ~level ~message:line;
                      Buffer.add_char buf '\n';
                      Uwt_io.write oc (Buffer.contents buf))
                   lines >>= fun () ->
                 Uwt_io.flush oc
               end channel)
    ~close:(match close_mode with
              | `Keep -> Lwt.return
              | `Close -> (fun () -> Uwt_io.close channel))

let () = Lwt_log_core.default := channel ~close_mode:`Keep ~channel:Uwt_io.stderr ()

let file ?(template="$(date): $(section): $(message)") ?(mode=`Append) ?(perm=0o640) ~file_name () =
  let mode = match mode with
    | `Append ->
        [Uwt.Fs.O_WRONLY; Uwt.Fs.O_CREAT; Uwt.Fs.O_APPEND; Uwt.Fs.O_NONBLOCK]
    | `Truncate ->
        [Uwt.Fs.O_WRONLY; Uwt.Fs.O_CREAT; Uwt.Fs.O_TRUNC; Uwt.Fs.O_NONBLOCK] in
  Uwt.Fs.openfile ~perm ~mode file_name >>= fun fd ->
  let oc = Uwt_io.of_file ~mode:Uwt_io.output fd in
  Lwt.return (channel ~template ~close_mode:`Close ~channel:oc ())

let level_code = function
  | Lwt_log_core.Fatal -> 0
  | Lwt_log_core.Error -> 3
  | Lwt_log_core.Warning -> 4
  | Lwt_log_core.Notice -> 5
  | Lwt_log_core.Info -> 6
  | Lwt_log_core.Debug -> 7

type syslog_facility =
    [ `Auth | `Authpriv | `Cron | `Daemon | `FTP | `Kernel
    | `Local0 | `Local1 | `Local2 | `Local3 | `Local4 | `Local5 | `Local6 | `Local7
    | `LPR | `Mail | `News | `Syslog | `User | `UUCP | `NTP | `Security | `Console ]

let facility_code = function
  | `Kernel -> 0
  | `User -> 1
  | `Mail -> 2
  | `Daemon -> 3
  | `Auth -> 4
  | `Syslog -> 5
  | `LPR -> 6
  | `News -> 7
  | `UUCP -> 8
  | `Cron -> 9
  | `Authpriv -> 10
  | `FTP -> 11
  | `NTP -> 12
  | `Security -> 13
  | `Console -> 14
  | `Local0 -> 16
  | `Local1 -> 17
  | `Local2 -> 18
  | `Local3 -> 19
  | `Local4 -> 20
  | `Local5 -> 21
  | `Local6 -> 22
  | `Local7 -> 23

type syslog_connection_type = STREAM | DGRAM

let shutdown fd =
  Uwt.Pipe.shutdown fd >>= fun () ->
  Uwt.Pipe.close_noerr fd;
  Lwt.return_unit

(* Try to find a socket in [paths]. For each path it check that the
   file is a socket and try to establish connection in DGRAM mode then in
   STREAM mode. *)
let syslog_connect paths =
  let rec loop = function
    | [] ->
        (* No working socket found *)
        log_intern "no working socket found in {%s}; is syslogd running?"
          (String.concat ", " (List.map (Printf.sprintf "\"%s\"") paths));
        Lwt.fail (Sys_error(Unix.error_message Unix.ENOENT))
    | path :: paths ->
        begin try
          Lwt.return (Some (Unix.stat path).Unix.st_kind)
        with
          | Unix.Unix_error(Unix.ENOENT, _, _) ->
              Lwt.return_none
          | Unix.Unix_error(error, _, _) ->
              log_intern "can not stat \"%s\": %s" path (Unix.error_message error);
              Lwt.return_none
        end >>= function
          | None ->
              loop paths
          | Some Unix.S_SOCK ->
            Lwt.catch ( fun () ->
              let fd = Unix.socket Unix.PF_UNIX Unix.SOCK_DGRAM 0 in
              Lwt.catch ( fun () ->
                  Unix.connect fd (Unix.ADDR_UNIX path);
                  Unix.set_close_on_exec fd;
                  Unix.set_nonblock fd;
                  let p = Uwt.Pipe.openpipe_exn ~ipc:false fd in
                  Lwt.return (DGRAM, p)
                )
                (function
                | Unix.Unix_error(Unix.EPROTOTYPE, _, _) ->
                  (* Then try with a stream socket: *)
                  let fd = Unix.socket Unix.PF_UNIX Unix.SOCK_STREAM 0 in
                  Lwt.catch
                    (fun () ->
                       Unix.connect fd (Unix.ADDR_UNIX path);
                       Unix.set_close_on_exec fd;
                       Unix.set_nonblock fd;
                       let p = Uwt.Pipe.openpipe_exn ~ipc:false fd in
                       Lwt.return (STREAM, p)
                    )
                    (function
                      | Unix.Unix_error(error, _, _) ->
                        Unix.close fd;
                        log_intern "can not connect to \"%s\": %s" path (Unix.error_message error);
                        loop paths
                      | exn -> Lwt.fail exn)
                | Unix.Unix_error(error, _, _) ->
                    Unix.close fd ;
                    log_intern "can not connect to \"%s\": %s" path (Unix.error_message error);
                    loop paths
                | exn -> Lwt.fail exn)
              ) ( function x -> Lwt.fail x)
          | Some _ ->
              log_intern "\"%s\" is not a socket" path;
              loop paths
  in
  loop paths

(* Write the whole contents of a string on the given file
   descriptor: *)
let write_string fd str =
  Uwt.Pipe.write_string fd ~buf:str

let truncate buf max =
  if Buffer.length buf > max then begin
    let suffix = "<truncated>" in
    let len_suffix = String.length suffix in
    Buffer.sub buf 0 (max - len_suffix) ^ suffix
  end else
    Buffer.contents buf

let syslog ?(template="$(date) $(name)[$(pid)]: $(section): $(message)") ?(paths=["/dev/log"; "/var/run/log"; "/var/run/syslog"]) ~facility () =
  let syslog_socket = ref None and mutex = Lwt_mutex.create () in
  let get_syslog () = match !syslog_socket with
    | Some x ->
        Lwt.return x
    | None ->
        syslog_connect paths >>= fun x ->
        syslog_socket := Some x;
        Lwt.return x
  in
  make
    ~output:(fun section level lines ->
               Lwt_mutex.with_lock mutex
                 (fun () ->
                    let buf = Buffer.create 42 in
                    let make_line socket_type msg =
                      Buffer.clear buf;
                      Printf.bprintf buf "<%d>" ((facility_code facility lsl 3) lor level_code level);
                      render ~buffer:buf ~template ~section ~level ~message:msg;
                      if socket_type = STREAM then Buffer.add_char buf '\x00';
                      truncate buf 1024
                    in
                    let rec print socket_type fd = function
                      | [] ->
                          Lwt.return_unit
                      | line :: lines ->
                          Lwt.catch
                            (fun () ->
                              write_string fd (make_line socket_type line) >>= fun () ->
                              print socket_type fd lines)
                            (function
                            | Unix.Unix_error(_, _, _) ->
                                (* Try to reconnect *)
                                shutdown fd >>= fun () ->
                                syslog_socket := None;
                                get_syslog () >>= fun (socket_type, fd) ->
                                write_string fd (make_line socket_type line) >>= fun () ->
                                print socket_type fd lines
                            | exn -> Lwt.fail exn)
                    in
                    get_syslog () >>= fun (socket_type, fd) ->
                    print socket_type fd lines))
    ~close:(fun () ->
              match !syslog_socket with
                | None ->
                    Lwt.return_unit
                | Some(_socket_type, fd) ->
                    shutdown fd)
